pkg/pipeline/stream/parallel.go

```diff
@@ -0,0 +1,337 @@
+	// it has a buffer, so it can store one job for each worker
+	coordinatorJobs := make(chan parallelNodeJob, n.Workers)
+	var coordinatorWg sync.WaitGroup
+	coordinatorWg.Add(1)
+	coordinator := newParallelNodeCoordinator(n.ID(), coordinatorJobs, errs, n.logger, n.base.Send, donePool)
+	go func() {
+		defer coordinatorWg.Done()
+		coordinator.Run(ctx)
+	}()
+
+	// workersDone is closed once all workers stop running (for whatever reason)
+	workersDone := make(chan struct{})
+	go func() {
+		workerWg.Wait()
+		close(workersDone)
+	}()
+
+	defer func() {
+		close(workerJobs)
+		close(coordinatorJobs)
+		workerWg.Wait()
+		coordinatorWg.Wait()
+		for {
+			select {
+			case workerErr := <-errs:
+				err = cerrors.LogOrReplace(err, workerErr, func() {
+					n.logger.Warn(ctx).Err(workerErr).Msg("parallel worker node failed")
+				})
+			default:
+				return
+			}
+		}
+	}()
+
+	for {
+		msg, err := trigger()
+		if err != nil || msg == nil {
+			return err
+		}
+
+		job := parallelNodeJob{
+			Message: msg,
+			Done:    donePool.Get().(chan struct{}),
+		}
+
+		// try sending the job to a worker
+		select {
+		case workerJobs <- job:
+			// we submitted the job to a worker, give it to the coordinator as well
+			coordinatorJobs <- job
+		case <-workersDone:
+			// no worker is running anymore, they must have all failed, nack the
+			// message and stop running
+			noWorkerRunningErr := cerrors.New("no worker is running")
+			err = msg.Nack(noWorkerRunningErr, n.ID())
+			if err != nil {
+				return err
+			}
+			return noWorkerRunningErr
+		}
+	}
+}
+
+func (n *ParallelNode) Sub(in <-chan *Message) {
+	n.base.Sub(in)
+}
+
+func (n *ParallelNode) Pub() <-chan *Message {
+	return n.base.Pub()
+}
+
+func (n *ParallelNode) SetLogger(logger log.CtxLogger) {
+	n.logger = logger
+}
+
+type parallelNodeJob struct {
+	Message *Message
+	Done    chan struct{}
+}
+
+// parallelNodeCoordinator coordinates the messages that are processed by
+// workers and ensures their order.
+type parallelNodeCoordinator struct {
+	name     string
+	jobs     <-chan parallelNodeJob
+	errs     chan<- error
+	logger   log.CtxLogger
+	send     func(ctx context.Context, logger log.CtxLogger, msg *Message) error
+	donePool *sync.Pool
+}
+
+func newParallelNodeCoordinator(
+	name string,
+	jobs <-chan parallelNodeJob,
+	errs chan<- error,
+	logger log.CtxLogger,
+	send func(ctx context.Context, logger log.CtxLogger, msg *Message) error,
+	donePool *sync.Pool,
+) *parallelNodeCoordinator {
+	logger.Logger = logger.With().Str(log.ParallelWorkerIDField, name+"-coordinator").Logger()
+	return &parallelNodeCoordinator{
+		name:     name,
+		jobs:     jobs,
+		errs:     errs,
+		logger:   logger,
+		send:     send,
+		donePool: donePool,
+	}
+}
+
+func (c *parallelNodeCoordinator) Run(ctx context.Context) {
+	// fail toggles a short circuit that causes all future messages to be nacked
+	fail := false
+	for job := range c.jobs {
+		// wait for job to be done
+		<-job.Done
```

Comment:
IIUC, this is how message order is being maintained? Every message becomes a job, as they are coming, they are pushed into the jobs "queue" (i.e. channel) and taken out in the same order.

Reply:
Exactly 👍 I'll add a comment to make this clear.